google/generativeai/types/content_types.py

```diff
@@ -240,15 +251,336 @@ def to_contents(contents: ContentsType) -> list[glm.Content]:
+    elif callable(fun):
+        return CallableFunctionDeclaration.from_function(fun)
+    else:
+        raise TypeError(
+            "Expected an instance of `genai.FunctionDeclaraionType`. Got a:\n" f"  {type(fun)=}\n",
+            fun,
+        )
+
+
+def _encode_fd(fd: FunctionDeclaration | glm.FunctionDeclaration) -> glm.FunctionDeclaration:
+    if isinstance(fd, glm.FunctionDeclaration):
+        return fd
+
+    return fd.to_proto()
+
+
+class Tool:
+    def __init__(self, function_declarations: Iterable[FunctionDeclarationType]):
+        # The main path doesn't use this but is seems useful.
+        self._function_declarations = [_make_function_declaration(f) for f in function_declarations]
+        self._index = {}
+        for fd in self._function_declarations:
+            name = fd.name
+            if name in self._index:
+                raise ValueError("")
+            self._index[fd.name] = fd
+
+        self._proto = glm.Tool(
+            function_declarations=[_encode_fd(fd) for fd in self._function_declarations]
+        )
+
+    @property
+    def function_declarations(self) -> list[FunctionDeclaration | glm.FunctionDeclaration]:
+        return self._function_declarations
+
+    def __getitem__(
+        self, name: str | glm.FunctionCall
+    ) -> FunctionDeclaration | glm.FunctionDeclaration:
+        if not isinstance(name, str):
+            name = name.name
+
+        return self._index[name]
+
+    def __call__(self, fc: glm.FunctionCall) -> glm.FunctionResponse | None:
+        declaration = self[fc]
+        if not callable(declaration):
+            return None
+
+        return declaration(fc)
+
+    def to_proto(self):
+        return self._proto
+
+
+class ToolDict(TypedDict):
+    function_declarations: list[FunctionDeclarationType]
+
+
+ToolType = Union[
+    Tool, glm.Tool, ToolDict, Iterable[FunctionDeclarationType], FunctionDeclarationType
+]
+
+
+def _make_tool(tool: ToolType) -> Tool:
+    if isinstance(tool, Tool):
+        return tool
+    elif isinstance(tool, glm.Tool):
+        return Tool(function_declarations=tool.function_declarations)
+    elif isinstance(tool, dict):
+        return Tool(**tool)
+    elif isinstance(tool, Iterable):
+        return Tool(function_declarations=tool)
+    else:
+        try:
+            return Tool(function_declarations=[_make_function_declaration(tool)])
+        except Exception as e:
+            raise TypeError(
+                "Expected an instance of `genai.ToolType`. Got a:\n" f"  {type(tool)=}",
+                tool,
+            ) from e
+
+
+class FunctionLibrary:
+
+    def __init__(self, tools: Iterable[ToolType]):
+        tools = _make_tools(tools)
+        self._tools = list(tools)
+        self._index = {}
+        for tool in self._tools:
+            for declaration in tool.function_declarations:
+                name = declaration.name
+                if name in self._index:
+                    raise ValueError("")
```

Comment:
```suggestion
                    raise ValueError(f"Tool named {name} already defined. Tools must be uniquely named.")
```

This seems like it could float up to users